### PR TITLE
fix: let non_canonical_impls skip proc marco

### LINKED ISSUE #12788: `non_canonical_clone_impl` should skip proc macros
### Summary

```
warning: non-canonical implementation of `clone` on a `Copy` type
   --> compiler/rustc_type_ir/src/ty_kind.rs:797:10
    |
797 | #[derive(derivative::Derivative)]
    |          ^^^^^^^^^^^^^^^^^^^^^^ in this derive macro expansion
    |
   ::: /home/matthias/.cargo/registry/src/index.crates.io-6f17d22bba15001f/derivative-2.2.0/src/lib.rs:77:1
    |
77  | pub fn derivative(input: TokenStream) -> TokenStream {
    | ---------------------------------------------------- in this expansion of `#[derive(derivative::Derivative)]`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#non_canonical_clone_impl
```
I guess there is not tooo much that can be one here? :sweat_smile: 

### Lint Name

non_canonical_clone_impl

### Reproducer

_No response_

### Version

_No response_

### Additional Labels

_No response_

### PR DESCRIPTION
**Pull request by WeiTheShinobi**
Fixed #12788

Although the issue only mentions `NON_CANONICAL_CLONE_IMPL`, this fix will also affect `NON_CANONICAL_PARTIAL_ORD_IMPL` because I saw
> Because of these unforeseeable or unstable behaviors, macro expansion should often not be regarded as a part of the stable API.

on Clippy Documentation and these two lints are similar, so I think it might be good, not sure if it's right or not.

--- 

changelog: `NON_CANONICAL_CLONE_IMPL`, `NON_CANONICAL_PARTIAL_ORD_IMPL` will skip proc marco now

### CODE REVIEW

File: tests/ui/non_canonical_clone_impl.fixed [hunk 1]
```diff
@@ -95,3 +96,7 @@ impl<A: Copy> Clone for Uwu<A> {
 }
 
 impl<A: std::fmt::Debug + Copy + Clone> Copy for Uwu<A> {}
+
+// should skip proc macros, see https://github.com/rust-lang/rust-clippy/issues/12788
+#[derive(proc_macro_derive::NonCanonicalClone)]
+pub struct G;
```
Comment (y21) [inline tests/ui/non_canonical_clone_impl.fixed:102]: While we're here, please also add a test case using `proc_macros::with_span` [like here](https://github.com/rust-lang/rust-clippy/blob/master/tests/ui/as_conversions.rs#L17). That's another common FP case.

Might need to add another check using `is_from_proc_macro` to the lint.

Comment (y21) [reply tests/ui/non_canonical_clone_impl.fixed:105]: Sorry, what I meant was something like like
```rs
with_span!(
    span

    #[derive(Copy)]
    struct H;
    impl Clone for H {
        fn clone(&self) -> Self {
            todo!()
        }
    }
);
```
(not using the proc macro) ^^

Comment (WeiTheShinobi) [reply tests/ui/non_canonical_clone_impl.fixed]: Sorry, it's my fault.
Thank you so much.

File: tests/ui/non_canonical_clone_impl.stderr [hunk 2]
```diff
@@ -29,13 +29,19 @@ LL | |     }
...
 LL | |         *self = source.clone();
 LL | |     }
    | |_____^ help: remove it
 
-error: aborting due to 4 previous errors
+error: non-canonical implementation of `clone` on a `Copy` type
+  --> tests/ui/non_canonical_clone_impl.rs:120:5
+   |
+LL |     span
+   |     ^^^^ help: change this to: `{ *self }`
```
Comment (y21) [inline tests/ui/non_canonical_clone_impl.stderr:44]: We shouldn't warn here. Most likely you need to change this to an `||`

https://github.com/rust-lang/rust-clippy/blob/64be449ad76fc9cb614e11ca9f758bc0c64b30d4/clippy_lints/src/non_canonical_impls.rs#L132

**Review by y21 — COMMENTED** [review]
Should be good to go with this one fixed

**Review by y21 — APPROVED** [review]

**Comment by y21** [comment]
Looks good, thanks! @bors r+

**Pull request merged**